CASSANDRA-16725 - implement nodetool getauditlog

smiklosovic opened the pull request.

michaelsembwever commented on src/java/org/apache/cassandra/audit/AuditLogOptions.java:
```diff
@@ -40,8 +47,220 @@
      */
-    public String audit_logs_dir = System.getProperty("cassandra.logdir.audit",
-                                                      System.getProperty("cassandra.logdir",".")+"/audit/");
+    public String audit_logs_dir = Paths.get(System.getProperty("cassandra.logdir.audit",
```
we shouldn't be specifying property keys directly anymore. please use `CassandraRelevantProperties`

smiklosovic replied:
good catch, fixed.

michaelsembwever commented on src/java/org/apache/cassandra/audit/AuditLogOptionsCompositeData.java:
```diff
@@ -0,0 +1,162 @@
+                                                            "excluded categories",
+                                                            "included users",
+                                                            "excluded users",
+                                                            "audit logger implementation class name" };
```
Rather than working with three arrays, could we introduce a class like
```
class AuditLogOption
{
    final String name;
    final String description;
    final SimpleType type;
}
```

smiklosovic replied:
I do not want to do this differently, actually. This is done like it is here across the whole project, I just do not want to introduce just another way how we do it, it is just good enough how it is, in my opinion.

michaelsembwever replied:
The suggestion is trivial, and was offered based on providing a) improved readability, and b) safety (if items are added).

(a) certainly weighs against what the precedence in the codebase already is. Might you provide examples of the precedence that is "across the whole project", as the use of multiple separate arrays being used consistently instead of simple structs surprises me a little?

smiklosovic replied:
I ll find some cases like that for you on Monday, however I can indeed do that if we do not mind the inconsistency. I have to admit I am more often than not doing something "as it was always done" just to not increase the "entropy" unnecessarilly.

michaelsembwever commented on src/java/org/apache/cassandra/audit/AuditLogOptionsCompositeData.java:
```diff
@@ -0,0 +1,162 @@
+
+import org.apache.cassandra.config.ParameterizedClass;
+
+public class AuditLogOptionsCompositeData
```
i'm thinking this class would be a better fit as an inner class to `GetAuditLog` since it is about a jmx represented view of the AuditLogOptions (i.e. let's keep the jmx specifics out of this package)

smiklosovic replied:
While I understand what you mean, it looks weird to depend in StorageService on some stuff from the nodetool package.

michaelsembwever replied:
>  it looks weird to depend in StorageService on some stuff from the nodetool package.

Good point.

So two things here now come to mind…
1) why are we putting `getAuditLogOptions()` into StorageService? Would it be cleaner we a `AuditLogManagerMBean` ?
2) how do we code this to make it easy for the virtual table to be added?

WRT (2) the codebase is going to find new precedences to accommodate the transition period of JMX and Virtual Tables, and it would be nice if each didn't involve a verbose copy of code. (I'm not thinking about any high level compatibility layer as was discussed on the ML, just code with a small amount of effort that makes sense.)

If we have time (give me a week or two), I wouldn't mind offering up a patch for the virtual tables addition to this. That would help see if there's anything to (2).

smiklosovic replied:
That is actually a good idea to create separate MBean for that. I might do that.

I really appreciate your willingness to help here in regards of upcoming changes related to virtual tables and similar, however, honestly, and I do not want to look like I am rushing this, but just for the sake of the simplicity and getting rid of this, I would go with the easy path, not trying to come up with anything new, until we prepare CEP where all these things will be written down more clearly. I am afraid that anything we come up with here will be rewritten anyway when that virtual tables CEP arrives.

michaelsembwever replied:
There won't be a CEP for virtual tables, they already exist, and we are free to add virtual tables simply under improvement tickets. 

My suggestion and offer to write up a quick virtual table was not for inclusion in this patch (though it can be if you want it, I was envisioning it as a subsequent jira ticket), but for a quick evaluation of the technical debt possibly avoided. Let me offer it up and I can then better explain.

michaelsembwever replied:
https://github.com/instaclustr/cassandra/compare/CASSANDRA-16725...thelastpickle:mck/instaclustr/CASSANDRA-16725/trunk

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -148,6 +154,8 @@ public synchronized void disableAuditLog()
         unregisterAsListener();
         IAuditLogger oldLogger = auditLogger;
         auditLogger = new NoOpAuditLogger(Collections.emptyMap());
+        // when we disable audit logging, we should also reset options so we return default (from cassandra.yml)
```
```suggestion
        // when we disable audit logging, we should also reset options so we return default ones (from cassandra.yaml)
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -177,6 +195,16 @@ public synchronized void enable(AuditLogOptions auditLogOptions) throws Configur
         oldLogger.stop();
     }
 
+    private void updateAuditOptions(final AuditLogOptions options, final AuditLogFilter filter)
```
```suggestion
    private void updateAuditLogOptions(final AuditLogOptions options, final AuditLogFilter filter)
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogFilter.java:
```diff
@@ -31,12 +30,12 @@
+    public final ImmutableSet<String> excludedCategories;
+    public final ImmutableSet<String> includedCategories;
+    public final ImmutableSet<String> includedUsers;
+    public final ImmutableSet<String> excludedUsers;
```
I am not sure it is a good idea to directly make those `public `. If we make a change, let it be at least to `protected`

smiklosovic replied:
I made it _package protected_ as well as the whole class (plus I made it final). It does not need to be protected when that class is never going to be extended anway.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/config/CassandraRelevantProperties.java:
```diff
@@ -159,6 +159,10 @@
+    LOG_DIR("cassandra.logdir", "."),
+
+    LOG_DIR_AUDIT("cassandra.logdir.audit"),
+
```
I suggest we add a java doc for those two, similar to other properties.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -158,16 +179,26 @@ public synchronized void disableAuditLog()
-        // always reload the filters
-        filter = AuditLogFilter.create(auditLogOptions);
-
-        // next, check to see if we're changing the logging implementation; if not, keep the same instance and bail.
-        // note: auditLogger should never be null
```
Shouldn't we keep this comment?

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tools/NodeProbe.java:
```diff
@@ -134,6 +138,7 @@
     protected HintsServiceMBean hsProxy;
     protected BatchlogManagerMBean bmProxy;
     protected ActiveRepairServiceMBean arsProxy;
+    protected AuditLogManagerMBean aloProxy;
```
I guess you wanted to type `almProxy`, not `aloProxy`

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -5786,35 +5786,40 @@ public void disableAuditLog()
-        loggerName = loggerName != null ? loggerName : DatabaseDescriptor.getAuditLoggingOptions().logger.class_name;
-
-        Preconditions.checkNotNull(loggerName, "cassandra.yaml did not have logger in audit_logging_option and not set as parameter");
-        Preconditions.checkState(FBUtilities.isAuditLoggerClassExists(loggerName), "Unable to find AuditLogger class: "+loggerName);
```
Are you sure you don't want these two preconditions anymore?

smiklosovic replied:
it will default to BinAuditLogger (default implementation) when this is not set and if it is set but the class does not exist, it will throw.

michaelsembwever commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -3139,7 +3149,7 @@ public static AuditLogOptions getAuditLoggingOptions()
     public static void setAuditLoggingOptions(AuditLogOptions auditLoggingOptions)
     {
-        conf.audit_logging_options = auditLoggingOptions;
+        conf.audit_logging_options = new AuditLogOptions.Builder(auditLoggingOptions).build();
```
nit: this line should be the only change in this file. Can all the import statement changes be reverted please.

smiklosovic replied:
thanks, fixed locally, will push it sometimes

michaelsembwever commented on src/java/org/apache/cassandra/db/virtual/AuditLogOptionsTable.java:
```diff
@@ -0,0 +1,100 @@
+    private static final String EXCLUDED_CATEGORIES = "excluded_categories";
+    private static final String INCLUDED_USERS = "included_users";
+    private static final String EXCLUDED_USERS = "excluded_users";
+    private static final String LOGGER = "logger";
```
curiosity: before we remove this vtable… if we move these constants into `AuditLogOptions`, would this remove the duplication of this string literals in this class and in `AuditLogOptionsCompositeData.AuditLogOption` ? 

(this was the primary purpose of raising during this review the vtable code.)

smiklosovic replied:
yes, I think we can do that, so once we introduce this we will just reuse them in this constructor.

smiklosovic committed:
implement getauditlog command

patch by Stefan Miklosovic; reviewed by Ekaterina Dimitrova, Mick Semb Wever for CASSANDRA-16725

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/tools/nodetool/GetAuditLogTest.java:
```diff
@@ -0,0 +1,142 @@
+
+        disableAuditLog();
+        testDefaultOutput(getAuditLog());
+    }
```
I feel we have four tests combined in one, can we split to make it easier when we debug tests? 
I appreciate the consistency with the `GetFullQueryLogTest` but I think they had only two tests there.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/audit/AuditLogFilterTest.java:
```diff
@@ -28,6 +28,22 @@
+        AuditLogOptions auditLogOptions = new AuditLogOptions.Builder()
+            .withIncludedKeyspaces(" ks, ks1, ks3, ")
+            .withEnabled(true)
+            .build();
```
nit: alignment

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tools/NodeTool.java:
```diff
@@ -209,6 +209,7 @@ public int execute(String... args)
                 RelocateSSTables.class,
                 ViewBuildStatus.class,
                 ReloadSslCertificates.class,
+                GetAuditLog.class,
```
I think this is fine where it is but at the same time get and set methods are grouped together and all other Getters are in a separate group L127-139. Maybe we should move it there? WDYT?

ekaterinadimitrova2 reviewed (commented):
Not a full review, but some initial comments. I need to spend more time to think about the MBean and a few other things. Just wanted to send these as I see in the meanwhile the patch is changing and every time I get back to it there are new commits and discussions so I am leaving what I have for now here for visibility and I will continue later. I also agree with the points made from Mick.

ekaterinadimitrova2 commented on test/unit/org/apache/cassandra/config/DatabaseDescriptorRefTest.java:
```diff
@@ -54,6 +54,8 @@
 public class DatabaseDescriptorRefTest
 {
     static final String[] validClasses = {
+    "org.apache.cassandra.config.CassandraRelevantProperties",
+    "org.apache.cassandra.config.CassandraRelevantProperties$PropertyConverter",
```
Those should be moved down I think, I see the classes are listed in alphabetical order

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -148,6 +167,8 @@ public synchronized void disableAuditLog()
         unregisterAsListener();
         IAuditLogger oldLogger = auditLogger;
         auditLogger = new NoOpAuditLogger(Collections.emptyMap());
+        // when we disable audit logging, we should also reset options so we return default ones (from cassandra.yml)
```
```suggestion
        // when we disable audit logging, we should also reset options so we return default ones (from cassandra.yaml)
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -148,6 +167,8 @@ public synchronized void disableAuditLog()
         IAuditLogger oldLogger = auditLogger;
         auditLogger = new NoOpAuditLogger(Collections.emptyMap());
+        // when we disable audit logging, we should also reset options so we return default ones (from cassandra.yml)
+        auditLogOptions = DatabaseDescriptor.getAuditLoggingOptions();
```
Should we really return to default ones? 
CC  @michaelsembwever 
I feel we need to explicitly notify the users that their config (if they have provided during enable from nodetool different one than the one in cassandra.yaml) will be reset. And probably also give them the chance to save the one they have? 
I understand that this is not change of behavior as before they didn't have the options configurable from nodetool but now when they have them, I think we need to let them know what to expect explicitly. What do you think?

smiklosovic replied:
I think I am following the same behavior how it is done in fql. I was not changing that.

smiklosovic replied:
I am taking my last response back. I ll return to this later.

ekaterinadimitrova2 replied:
Thanks, I also realized that actually the log dir was configurable before and there was no reset for it so there is a change of behavior unfortunately. Thanks for looking into that

smiklosovic replied:
I think that for some case (not sure if it is fql or audit, I dont remeber now), I _think_ that it was a plan to be _not able_ to change the log dir. Think about that, if somebody hacks jmx, for example, he might change the destination full query logs would be written - where he is able to read them. So that was done not settable on purpose I think.

ekaterinadimitrova2 replied:
You are right about the log dir, I misread. It is even in the docs:
https://cassandra.apache.org/doc/latest/cassandra/new/auditlogging.html#enabling-audit-logging-with-nodetool

ekaterinadimitrova2 replied:
But still we need to think about this reset how to make the users aware of it so there is no confusion; if we decide to really keep the reset. I don't think we reset the options for FQL? Do we?

smiklosovic replied:
@ekaterinadimitrova2 FQL is returning options from cassandra.yaml when FQL is disabled (1).

Here I would expect similar behavior, when I disable audit log, it should return options from cassandra.yaml.

If we want to do this in such a way that it would "persist" these options from the last time it was enabled, that is something else. I would create new ticket for this and we would cover both cases at once instead of doing it just here and then FQL would be inconsistent with audit log behaviour.

As of now, it is explicitly mentioned what they can expect to be returned in respective nodetool's command description.

(1) https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/fql/FullQueryLogger.java#L123-L145

ekaterinadimitrova2 replied:
I might have forgotten something as it's been a week since I was looking into the details but if we read directly from the cassandra.yaml, why do we have to do reset when we disable? And next time when we enable it just loads the default ones on enable? I see that is what the fql tool does. What do I miss? The stop method of the fqltool does not do any reset.

smiklosovic replied:
I dont know what you want me to say, honestly :) It is just done like that? We can change it but thats the scope of the other ticket how I wrote.

smiklosovic replied:
One more thing related to this change, it is harder to explain here, you would need to see the code so take a deeper look please, but conceptually, I might just set "options.enabled = false" upon disabling so all other options would stay there and if you enable again, we would just take these options and I would set "options.enabled = true" etc ...

but you might (re)enable your audit logger differently next time, if  you enable and disable it, if you want to enable it again, it does not necessarilly mean that all your options from the last enablement time are OK for you this time, you might change them.

So we would have to "merge" the old options with the new options in order to rewrite these which changed, but keep these which are same.

I think this just adds additional complexity and it is quite confusing for users. I prefer to "enable" and "disable" and _that's that_. I do not want to keep any "previous cached options". I just want to start with clean table next time I enable.

FQL behaves differently on the code level but in practice it behaves same.

However, if we decide we truly want to do this, I think that we need yet another method for "resetting options". Because if they are "remembered" every time then I guess you want to also "get factory settings" too.

But really, I would just keep that as is. Because, yet another example, there is StorageServiceMBean with enablement methods which accept a bunch of parameters. These methods are meant to be either called programmatically or via tooling like JConsole or what have you. Upon calling this method, you basically have to enumerate all of parameters. Now you can left some blank, so "null" would be propagated, but in that case - what value do you actually want to use? Does "null" mean that you want to "keep the old value you used the last time" or do you mean you want to trully use "null" in such sense that it would take parameters from cassandra.yaml?

I think we are making this just more complex then it should be.

ekaterinadimitrova2 replied:
I think I confused you. I understand both FQL tool and AuditLog take on enable the default cassandra.yaml parameters and whatever is provided command line arguments from NodeTool. I don't want to complicate it and change that.
My question was why we don't do:
```
public AuditLogOptions getAuditLogOptions()
    {
        if(isEnabled())
            return auditLogOptions;
        else
            // otherwise get what database is configured with from cassandra.yaml
            return  DatabaseDescriptor.getAuditLoggingOptions();
    }
```
This is how we do it for FQL log. This will make the code consistent and we don't need to do a reset in disable.

ekaterinadimitrova2 replied:
I will play a bit with the code now to see what I am missing but at first glance I don't see a reason to do it in a different way

ekaterinadimitrova2 replied:
I don't see a reason not to change the getAuditLogOptions to the way FQL does it. I think this will make it also more clear to people. Now when you look at getAuditLogOptions you will know exactly in what case what is returned. Reset in disable is not wrong(it obviously leads to the same behavior) but it is not self-explanatory.

smiklosovic replied:
ok I see now, I am glad we are on the same page, yeah lets explore that idea.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tools/nodetool/EnableAuditLog.java:
```diff
@@ -47,9 +49,37 @@
     @Option(title = "excluded_users", name = { "--excluded-users" }, description = "Comma separated list of users to be excluded for audit log. If not set the value from cassandra.yaml will be used")
     private String excluded_users = null;
 
+    @Option(title = "roll_cycle", name = {"--roll-cycle"}, description = "How often to roll the log file (MINUTELY, HOURLY, DAILY).")
```
Do we know why those were missing at first place?

smiklosovic replied:
I dont think we know this, probably they were just forgotten and not exposed, they are used in FQL. Audit logging uses same BinLog "framework" but we just havent exposed these parameters but FQL exposes them. I asked Marcus and he said "yeah why not".

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageServiceMBean.java:
```diff
@@ -790,8 +790,15 @@ public StageConcurrency(int corePoolSize, int maximumPoolSize)
+                               String includedUsers, String excludedUsers) throws ConfigurationException, IllegalStateException;
+
+    public void enableAuditLog(String loggerName, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
+                               String includedUsers, String excludedUsers) throws ConfigurationException, IllegalStateException;
```
So we don't use in the codebase this particular method anymore but we keep it for now? Do you know the reason why those options were not added initially to be configurable from NodeTool?

ekaterinadimitrova2 replied:
About the MBean... it is a good question.
I am not sure how do we act in those cases. Probably the options are two - we leave it "as is" for now and maybe follow up in additional ticket.
Or deprecate them with some backward compatibility? Similar to what we did with some metrics which names were changed. Also, I am not sure that this could be done in a minor version. (I mean breaking change shouldn't be in a minor version as far as I know) CC @michaelsembwever  as I feel he might have more experience with that than me. Mine is more of a personal thoughts than project experience

smiklosovic replied:
It was / is used in 4.0.0. But I kept it there because of the "compatibility". There is a rule I am following that there is some "unwritten contract" among devs (e.g. ask Markus Ericsson) that if we add a method we just can not remove the one which was there because we might disrupt the behaviour people rely on. So the common practice is to at least annotate it as obsolete (I can do that) and then it will be removed ... sometimes.

smiklosovic replied:
I deprecated methods which are not accepting new configuration options we are exposing (these low-level ones for bin log as it is done for fql)

I keep them there until the next release, after that I remove these deprecated methods and I move the ones which left in StorageService to new AuditLogManagerMBean.

If we want to keep this "deprecation logic" in next release when I am about to move them to dedicated MBean, I need to actually keep them in StorageService too as they just can not suddenly "disappear". We should deprecate the methods we move first and have them in dedicated MBean first. Hence I suspect there would be little bit of duplication as I do not think that StorageService might reference some other MBean.

ekaterinadimitrova2 replied:
I think that will be done with alias. For reference, I found this good example today https://issues.apache.org/jira/browse/CASSANDRA-16404?page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel&focusedCommentId=17400403#comment-17400403

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogOptions.java:
```diff
@@ -60,6 +300,7 @@ public String toString()
                ", block=" + block +
                ", max_queue_weight=" + max_queue_weight +
                ", max_log_size=" + max_log_size +
+               ", max_archive_retries=" + max_archive_retries +
```
Not sure whether this change might not break someone's parsing actually?

smiklosovic replied:
hardly, nobody parses this anyway ... I am fixing things here acutally :)

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogOptions.java:
```diff
@@ -37,11 +46,242 @@
+        {
+            this.auditLogDir = auditLogDir;
+            return this;
+        }
```
I just realized we don't use this one and I am wondering whether we expect to need It in the future. Not very strong opinion around that though. Maybe we can just suppress the warning for now:
```
@SuppressWarnings("unused")
public Builder withAuditLogDir(final String auditLogDir)
```
I leave it up to you.

smiklosovic replied:
let's keep it how it is

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -5786,38 +5786,53 @@ public void disableAuditLog()
         logger.info("Auditlog is disabled");
     }
 
+    @Deprecated
```
nit: I believe you forgot the corresponding @Deprecated in `StorageServiceMBean`.  This can be done on commit

smiklosovic replied:
ah right ...

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogManager.java:
```diff
@@ -83,16 +86,27 @@ public void initialize()
+    @Override
+    public CompositeData getAuditLogOptionsData()
+    {
+        return AuditLogOptionsCompositeData.toCompositeData(AuditLogManager.instance.getAuditLogOptions());
```
nit: And I leave it really to you to decide but I would probably put this method before or after` getAuditLogOptions`
The methods in between are short but this might change and I feel those two will be probably better together. Feel free to ignore me if you don't like the suggestion :-)

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogOptions.java:
```diff
@@ -37,11 +46,242 @@
+
+    public AuditLogOptions()
+    {
+        String auditLogDir = CassandraRelevantProperties.LOG_DIR_AUDIT.getString();
```
nit: I feel this should be `auditLogsDir` for code consistency and the below one `logsDir`. You can change it on commit if you want to

ekaterinadimitrova2 replied:
or `audit_logs_dir` can be changed to `audit_log_dir` as actually the property says cassandra.logdir. I leave it to you.

ekaterinadimitrova2 replied:
Yeah, looking around everywhere it is logDir so I suggest we change just ` audit_logs_dir` to `audit_log_dir` on commit for consistency

ekaterinadimitrova2 replied:
Ok and in the cassandra.yaml it is audit_logs_dir..... Just leave it as is. Sorry, it was done confusing initially with difference between cassandra.logdir but in Cassandra.yaml audit_logs_dir.... 
Anyway, I am happy as is.

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogOptions.java:
```diff
@@ -37,11 +46,242 @@
+        }
+
+        public Builder withEnabled(boolean enabled)
+        {
```
For the record: the BinLog builder is missing the `with` prefix but I think your naming convention is better 👍

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogOptionsCompositeData.java:
```diff
@@ -0,0 +1,232 @@
+           (opts, obj) -> opts.roll_cycle = (String) obj),
+
+    option(AuditLogOption.BLOCK,
+           "indicates if the BinLog should block if the it falls behind or should drop bin log records",
```
```suggestion
           "indicates if the BinLog should block if it falls behind or it should drop bin log records",
```

smiklosovic replied:
thanks

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/tools/nodetool/EnableAuditLog.java:
```diff
@@ -47,9 +49,37 @@
+    @Option(title = "max_log_size", name = {"--max-log-size"}, description = "How many bytes of log data to store before dropping segments. Might not be respected if a log file hasn't rolled so it can be deleted.")
+    private long maxLogSize = Long.MIN_VALUE;
+
+    @Option(title = "archive_command", name = {"--archive-command"}, description = "Command that will handle archiving rolled full query log files." +
```
```suggestion
    @Option(title = "archive_command", name = {"--archive-command"}, description = "Command that will handle archiving rolled audit log files." +
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/audit/AuditLogOptionsCompositeData.java:
```diff
@@ -0,0 +1,232 @@
+           (opts, obj) -> opts.block = (Boolean) obj),
+
+    option(AuditLogOption.MAX_QUEUE_WEIGHT,
+           "maximum weight of in memory queue for records waiting to be written to the binlog file before blocking or dropping the log records",
```
```suggestion
           "maximum weight of in-memory queue for records waiting to be written to the binlog file before blocking or dropping the log records",
```

ekaterinadimitrova2 commented on src/java/org/apache/cassandra/service/StorageService.java:
```diff
@@ -5786,38 +5786,53 @@ public void disableAuditLog()
     }
 
+    @Deprecated
     public void enableAuditLog(String loggerName, Map<String, String> parameters, String includedKeyspaces, String excludedKeyspaces, String includedCategories, String excludedCategories,
```
Hey, I just noticed the upper method, not this one is a non-used one and I guess that one had to be actually deprecated?

smiklosovic replied:
no no, that one is supposed to be used.

ekaterinadimitrova2 replied:
There is one unused enableAuditLog method that is not deprecated and I am not sure why, can you, please, double check?

smiklosovic replied:
that one does not accept Map, for tooling like JConsole, there is not any way how to invoke a method if it accepts Map, it is not invokable there.

smiklosovic replied:
People tend to use tooling different from nodetool and they are not calling that programmatically either, if there is just one method accepting paramter like Map, we just make it incallable in that tooling. JConsole, for example, is used quite heavily by our devops and they are more often in JConsole than nodetool (some of them) so I just do not want to cut them off completely here.

I know we might now try to figure out some workaround or similar but that would end up being even more awkward and custom and that is not the problem I want to address in this ticket.

Methods in interfaces are very inconsistent and they lack a robust and coherent approach and us spending any more time on this is rather quesionable ... when it comes to this ticket itself and what we try to achieve.

smiklosovic replied:
This specific method was deprecated because it does not contain all other parameters we enabled to be specified. The longer version is used instead.

ekaterinadimitrova2 replied:
Ok, so just add a comment on commit so that people don't have to ask, please. Thanks
